File: back-bardak/src/main/java/kz/bardak/game/runtime/MatchSession.java
package kz.bardak.game.runtime;

import java.util.LinkedHashSet;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.UUID;
import kz.bardak.game.rules.DealCommand;
import kz.bardak.game.rules.MatchEngine;
import kz.bardak.game.rules.MatchResult;
import kz.bardak.game.rules.MatchState;
import kz.bardak.game.rules.PlayerView;
import kz.bardak.game.rules.StateProjection;

/**
 * Идущий матч за одним столом.
 *
 * <p>⭐ Здесь же живёт единственное отображение «место за столом → место в движке».
 * Движок нумерует игроков подряд от нуля, а за столом места могут идти с пропусками —
 * кто-то встал до начала. Считать это отображение в двух местах означает рано или поздно
 * разойтись и посадить игрока не туда.
 *
 * <p>Объект не потокобезопасен намеренно: он живёт на потоке своего стола (ADR-007).
 */
public final class MatchSession {

    private final UUID tableId;
    private final UUID matchId;
    private final List<UUID> seatUsers;
    private final MatchEngine engine;
    private final StateProjection projection;

    private MatchState state;

    /** Сквозной номер события по матчу. Живёт на потоке стола, поэтому просто int. */
    private int lastSeq;

    /**
     * ⭐ Уже применённые команды.
     *
     * <p>Клиент переотправляет команду после разрыва — он не знает, дошла ли предыдущая.
     * Без дедупликации ход применился бы дважды: карта ушла бы со стола дважды или пас
     * закрыл бы чужой раунд. Помним последние {@value #REMEMBERED_COMMANDS} — этого
     * с запасом хватает на любую очередь переотправки, а расти бесконечно набору незачем.
     */
    private final LinkedHashSet<String> appliedCommands = new LinkedHashSet<>();

    public MatchSession(final UUID tableId, final UUID matchId, final List<UUID> seatUsers,
                        final MatchEngine engine, final StateProjection projection,
                        final MatchState state) {
        this.tableId = Objects.requireNonNull(tableId, "tableId");
        this.matchId = Objects.requireNonNull(matchId, "matchId");
        this.seatUsers = List.copyOf(Objects.requireNonNull(seatUsers, "seatUsers"));
        this.engine = Objects.requireNonNull(engine, "engine");
        this.projection = Objects.requireNonNull(projection, "projection");
        this.state = Objects.requireNonNull(state, "state");
    }

    private static final int REMEMBERED_COMMANDS = 200;

    public UUID tableId() {
        return tableId;
    }

    /** Команда с этим идентификатором уже применялась. */
    public boolean alreadyApplied(final String commandId) {
        return commandId != null && appliedCommands.contains(commandId);
    }

    public void remember(final String commandId) {
        if (commandId == null) {
            return;
        }
        appliedCommands.add(commandId);
        final var iterator = appliedCommands.iterator();
        while (appliedCommands.size() > REMEMBERED_COMMANDS && iterator.hasNext()) {
            iterator.next();
            iterator.remove();
        }
    }

    public UUID matchId() {
        return matchId;
    }

    public int lastSeq() {
        return lastSeq;
    }

    public int nextSeq() {
        return lastSeq + 1;
    }

    public void lastSeq(final int value) {
        this.lastSeq = value;
    }

    public MatchState state() {
        return state;
    }

    public List<UUID> players() {
        return seatUsers;
    }

    /** Место игрока в движке. Пусто — этот игрок за столом не играет (зашёл смотреть). */
    public Optional<Integer> seatOf(final UUID userId) {
        final int seat = seatUsers.indexOf(userId);
        return seat < 0 ? Optional.empty() : Optional.of(seat);
    }

    public UUID userAt(final int seatNo) {
        return seatUsers.get(seatNo);
    }

    /** Персональная проекция: чужих карт в ней нет физически (ADR-026). */
    public PlayerView viewFor(final UUID userId) {
        final int seat = seatOf(userId).orElseThrow(() ->
                new IllegalArgumentException("Игрок не за этим столом: " + userId));
        return projection.project(state.deal(), seat);
    }

    /** Применить команду. Состояние меняется, только если движок её принял. */
    public MatchResult apply(final DealCommand command) {
        final MatchResult result = engine.apply(state, command);
        if (result instanceof MatchResult.Applied applied) {
            state = applied.state();
        }
        return result;
    }
}

File: back-go/internal/application/match_session.go
package application

import (
	"fmt"
	"sync"

	"github.com/awesomeme01/bardak/back-go/internal/domain/game"
)

// MatchSession — живой матч за одним столом.
//
// ⚠️ Значение изменяемое и НЕ потокобезопасное по умолчанию: им владеет goroutine стола,
// и снаружи к нему обращаются только через её очередь. Мьютекс здесь всё же есть — но
// лишь для чтения состояния наблюдателями (снимок для нового подписчика), а не для игры.
type MatchSession struct {
	TableID string
	MatchID string

	// Seats — кто на каком месте; порядок зафиксирован на старте матча.
	//
	// ⚠️ Места берутся из МАТЧА, а не из лобби: лобби живёт своей жизнью, кто-то встал,
	// кто-то сел, а матч раздан по местам, зафиксированным при старте. Взять их из лобби
	// значит после перезапуска отдать игроку чужую руку.
	Seats []SeatOwner

	mu      sync.RWMutex
	state   game.MatchState
	lastSeq int

	engine  *game.MatchEngine
	config  game.RulesConfig
	applied *appliedCommands
}

// SeatOwner — кто сидит на месте.
type SeatOwner struct {
	SeatNo      int
	UserID      string
	DisplayName string
}

// NewMatchSession собирает сессию.
func NewMatchSession(tableID, matchID string, seats []SeatOwner, config game.RulesConfig,
	state game.MatchState) *MatchSession {
	return &MatchSession{
		TableID: tableID,
		MatchID: matchID,
		Seats:   seats,
		state:   state,
		engine:  game.NewMatchEngineFor(config),
		config:  config,
		applied: newAppliedCommands(),
	}
}

// State — текущее состояние матча.
func (s *MatchSession) State() game.MatchState {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.state
}

// LastSeq — номер последнего записанного события.
func (s *MatchSession) LastSeq() int {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.lastSeq
}

// SetLastSeq запоминает номер последнего события.
func (s *MatchSession) SetLastSeq(seq int) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.lastSeq = seq
}

// NextSeq — номер, который получит следующее событие.
func (s *MatchSession) NextSeq() int { return s.LastSeq() + 1 }

// Config — правила стола.
func (s *MatchSession) Config() game.RulesConfig { return s.config }

// SeatOf — место игрока; false, если он за этим столом не играет.
func (s *MatchSession) SeatOf(userID string) (int, bool) {
	for _, seat := range s.Seats {
		if seat.UserID == userID {
			return seat.SeatNo, true
		}
	}
	return 0, false
}

// Naming — как назвать место: кто там сидит и под каким именем.
func (s *MatchSession) Naming(seatNo int) (string, string) {
	for _, seat := range s.Seats {
		if seat.SeatNo == seatNo {
			return seat.UserID, seat.DisplayName
		}
	}
	// ⚠️ Место без владельца — не повод падать: так же ведёт себя Java, подставляя
	// прочерк. Матч важнее аккуратности подписи.
	return "", "—"
}

// Players — идентификаторы игроков матча, в порядке мест.
func (s *MatchSession) Players() []string {
	out := make([]string, 0, len(s.Seats))
	for _, seat := range s.Seats {
		out = append(out, seat.UserID)
	}
	return out
}

// Apply применяет команду к матчу.
//
// ⚠️ Ошибка движка — это не отказ игроку, а поломка: состояние в таком случае НЕ меняется,
// и вызывающий обязан отличать «ход недопустим» от «что-то сломалось».
func (s *MatchSession) Apply(command game.DealCommand) (game.MatchOutcome, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	outcome, err := s.engine.Apply(s.state, command)
	if err != nil {
		return game.MatchOutcome{}, err
	}
	if outcome.Applied {
		s.state = outcome.State
	}
	return outcome, nil
}

// IsOver — матч закончен.
func (s *MatchSession) IsOver() bool {
	return s.State().Phase == game.MatchOver
}

// AlreadyApplied — команда с таким идентификатором уже применялась.
func (s *MatchSession) AlreadyApplied(id string) bool { return s.applied.has(id) }

// Remember отмечает команду применённой.
//
// ⚠️ Зовётся ТОЛЬКО после успешного применения. Запомнить отклонённую значит на повтор
// вернуть снимок вместо причины отказа, и игрок не узнает, почему ход не прошёл.
func (s *MatchSession) Remember(id string) { s.applied.remember(id) }

// appliedCommands — окно применённых команд.
type appliedCommands struct {
	mu    sync.Mutex
	seen  map[string]struct{}
	order []string
}

const rememberedPerMatch = 200

func newAppliedCommands() *appliedCommands {
	return &appliedCommands{seen: make(map[string]struct{}, rememberedPerMatch)}
}

func (a *appliedCommands) has(id string) bool {
	if id == "" {
		return false
	}
	a.mu.Lock()
	defer a.mu.Unlock()
	_, ok := a.seen[id]
	return ok
}

func (a *appliedCommands) remember(id string) {
	if id == "" {
		return
	}
	a.mu.Lock()
	defer a.mu.Unlock()
	if _, ok := a.seen[id]; ok {
		return
	}
	a.seen[id] = struct{}{}
	a.order = append(a.order, id)
	if len(a.order) > rememberedPerMatch {
		delete(a.seen, a.order[0])
		a.order = a.order[1:]
	}
}

// ProjectFor — персональная проекция состояния для места.
func (s *MatchSession) ProjectFor(seatNo int) (game.PlayerView, error) {
	state := s.State()
	projection := game.NewStateProjection(s.config, game.NewDealEngineFor(s.config))
	view, err := projection.Project(state.Deal, seatNo)
	if err != nil {
		return game.PlayerView{}, fmt.Errorf("проекция для места %d: %w", seatNo, err)
	}
	return view, nil
}
